external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisRecordProcessor.scala

```diff
@@ -68,9 +69,16 @@ private[kinesis] class KinesisRecordProcessor[T](receiver: KinesisReceiver[T], w
   override def processRecords(batch: List[Record], checkpointer: IRecordProcessorCheckpointer) {
     if (!receiver.isStopped()) {
       try {
-        receiver.addRecords(shardId, batch)
-        logDebug(s"Stored: Worker $workerId stored ${batch.size} records for shardId $shardId")
-        receiver.setCheckpointer(shardId, checkpointer)
+        // Limit the number of processed records from Kinesis stream. This is because the KCL cannot
+        // control the number of aggregated records to be fetched even if we set `MaxRecords`
+        // in `KinesisClientLibConfiguration`. For example, if we set 10 to the number of max
+        // records in a worker and a producer aggregates two records into one message, the worker
+        // possibly 20 records every callback function called.
+        batch.asScala.grouped(receiver.getCurrentLimit).foreach { batch =>
+          receiver.addRecords(shardId, batch.asJava)
+          logDebug(s"Stored: Worker $workerId stored ${batch.size} records for shardId $shardId")
+          receiver.setCheckpointer(shardId, checkpointer)
```

Comment:
Yeah, that's what I suspected at https://github.com/apache/spark/pull/16114#discussion_r90756702 -- thanks for confirming